### PR TITLE
Smarter Zarf Registry HPA behavior

### PR DESCRIPTION
**Pull request by jeff-mccoy**
Make Registry HPA less aggressive and block from downscaling while running `zarf package deploy` if images exist.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging

- [x] Test, docs, adr added or updated as needed
- [x] [Contributor Guide Steps](https://github.com/defenseunicorns/zarf/blob/main/CONTRIBUTING.md#developer-workflow) followed

### CODE REVIEW

**Comment by jeff-mccoy** [comment]
During deploy, scaleDown disabled:
Screenshot 2023-02-01 at 7 20 27 PM

After deploy, scaleDown reset:
Screenshot 2023-02-01 at 7 20 33 PM

**Review by Racer159 — CHANGES_REQUESTED** [review]

File: src/pkg/packager/deploy.go [hunk 1]
```diff
@@ -68,6 +71,9 @@ func (p *Packager) Deploy() error {
 	// Note: Not all packages need k8s; check if k8s is being used before saving the secret
 	if p.cluster != nil {
 		p.cluster.RecordPackageDeployment(p.cfg.Pkg, deployedComponents)
+		if hpaModified {
```
[the remark this answers is not in the document]
Comment (jeff-mccoy) [reply src/pkg/packager/deploy.go:74]: no I need to address that, thx

File: src/pkg/packager/deploy.go [hunk 2]
```diff
@@ -216,6 +222,15 @@ func (p *Packager) deployComponent(component types.ZarfComponent, noImgChecksum
 			}
 		}
 
+		// Disable the registry HPA scale down if we are deploying images and it is not already disabled
+		if hasImages && !hpaModified {
+			if err := p.cluster.EnableRegistryHPAScaleDown(false); err != nil {
```
[the remark this answers is not in the document]
Comment (jeff-mccoy) [reply src/pkg/packager/deploy.go:227]: that's a really good point, will change

File: src/internal/cluster/zarf.go [hunk 3]
```diff
@@ -98,3 +99,30 @@ func (c *Cluster) RecordPackageDeployment(pkg types.ZarfPackage, components []ty
 
 	c.Kube.CreateOrUpdateSecret(deployedPackageSecret)
 }
+
+// EnableRegistryHPAScaleDown enables or disables the HPA scale down for the Zarf Registry.
+func (c *Cluster) EnableRegistryHPAScaleDown(enable bool) error {
```
Comment (YrrepNoj) [inline src/internal/cluster/zarf.go:104]: Nit; Having the function being named `EnableRegistryHPAScaleDown()` but then accepting a Boolean that affects if HPA is scaled down or not seems like an antipattern (or at least a bad name 😛)

Thoughts on changing the function header to something along the liens of `ConfigureRegistryHPAScaleDown(policy v2beta2.ScalingPolicySelect)` ?

**Review by YrrepNoj — DISMISSED** [review]
I have a single 'nit' over a function name.

Since it seems like just a nit, I'll approve the PR and leave resolving that to your judgement :)

File: src/internal/cluster/zarf.go (hunk 3 above)
Comment (Racer159) [reply src/internal/cluster/zarf.go]: @jeff-mccoy and I don't like the word "toggle".  That implies that it remembers state and flips back and forth when you call it.  I would prefer Set or Configure personally

Comment (wirewc) [reply src/internal/cluster/zarf.go]: @Racer159 I think toggle works great as most of the time you want it enabled to ensure a single core isn't getting hammered. Seeing that the issue originally reported by @corang shows that problem was an edge case for a specific deployment originally. Seeing the scope (as I understand it) is to temporarily change the behavior of HPA when deploying large packages, the word `toggle` seems accurate.

Comment (jeff-mccoy) [reply src/internal/cluster/zarf.go]: looked at different strategies and ended up just doing to diff functions. Not as DRY, but more readable I suppose...

`EnableRegHPAScaleDown()`
`DisableRegHPAScaleDown()`

Screenshot 2023-02-03 at 12 12 07 PM

File: src/internal/cluster/zarf.go [hunk 4]
```diff
@@ -98,3 +99,41 @@ func (c *Cluster) RecordPackageDeployment(pkg types.ZarfPackage, components []ty
 
 	c.Kube.CreateOrUpdateSecret(deployedPackageSecret)
 }
+
+// SetRegHPAScaleDown enables or disables the HPA scale down for the Zarf Registry.
+func (c *Cluster) EnableRegHPAScaleDown() error {
```
Comment (YrrepNoj) [inline src/internal/cluster/zarf.go:104]: The docstring comment title got lost as we were experimenting with function names

**Review by YrrepNoj — APPROVED** [review]

**Review by Racer159 — APPROVED** [review]
Lgtm

**Pull request merged**